### DISCUSSION TITLE
Cannot get local conflict resolved

### DISCUSSION THREAD

**Discussion by GordianDziwis** [tree-sitter/tree-sitter#1342] [Q&A] [answered]
I tried a lot. And I can not get it to work. The issue is that I have a local ambiguity and I do not know how to resolve it. 

I have a minimal grammar:
```javascript
module.exports = grammar({
  name: 'turtle',

  word: $ => $.pn_prefix,

  rules: {

    turtle_doc: $ => repeat($.statement),

    statement: $ => choice(
      $.sparql_base,
      seq($.namespace, '.')
    ),

    sparql_base: $ => seq(
      'BASE',
      token(seq( '<', /[A-Z]+/, '>')),
    ),

    namespace: $ => seq(
      optional($.pn_prefix),
      ':'
    ),

    pn_prefix: $ => /[A-Z]+/,
  }
})
```

And this:
```
BASE: .
```
should result in:
```
(turtle_doc (statement (namespace (pn_prefix)))) 
```
but it results in:
```
(turtle_doc (ERROR) (statement (namespace))) 
```

I tried prec, conflict token and dynamic prec with no avail. When I got the previous example parsed, this one broke:

```
BASE <EX>
BASE: .
```
```
(turtle_doc (statement (sparql_base)) (statement (namespace (pn_prefix))))
```

Full grammar is [here](https://github.com/BonaBeavis/tree-sitter-turtle/blob/main/grammar.js)

Can anyone help?

**Comment by tek** [comment]
did you try

```javascript
namespace: $ => seq(
  optional(choice('BASE', $.pn_prefix)),
  ':',
)
```

**Comment by ahlinc** [comment, marked as answer]
@BonaBeavis here is a full git diff with changes that you need _(all your tests passed)_:
<details><summary><i>git diff:</i></summary>

```diff
diff --git a/grammar.js b/grammar.js
index 5777f47..44757b9 100644
--- a/grammar.js
+++ b/grammar.js
@@ -82,12 +82,12 @@ const PN_LOCAL_ESC = [

 String.prototype.toCaseInsensitiv = function () {
   return alias(
-    token(new RegExp(
+    token(prec(1, new RegExp(
       this
         .split('')
         .map(letter => `[${letter}${letter.toLowerCase()}]`)
         .join('')
-    )),
+    ))),
     this
   )
 }
@@ -370,7 +370,10 @@ module.exports = grammar({

     // [139s]
     namespace: $ => seq(
-      optional($.pn_prefix),
+      optional(choice(
+        $.pn_prefix,
+        alias('BASE'.toCaseInsensitiv(), $.pn_prefix)
+      )),
       ':'
     ),
```

</details>

@tek previously pointed you into the right direction but there was a need to add a lexical precedence for your case insensitive `BASE` terminal cause it's declared as regexp. Also it seems you provided an incorrect `BASE: .` example cause by [the Turtle spec](https://www.w3.org/TR/turtle/#grammar-production-triples) it's not allowed to omit the `$.property_list`, for ex. for the `BASE: a [] .` it works correctly.

**Reply by GordianDziwis** [reply]
@ahlinc Thanks  a lot for your help. This opened a can of worms on edge cases. The general issue seems to be that the [productions for terminals](https://www.w3.org/TR/turtle/#grammar-production-triples) in the turtle grammar are very complex.
If I write for each terminal production one big token function it works. But this has several drawbacks, it is very difficult to read, there are a lot of repetitions and I need the nodes inside those token downstream.
If I compose those terminals out of multiple productions (like in the spec), I can't get the precedence right. String tokens from the grammar seem always have precedence and I can't use the word token because I did not define the terminal productions as a single token.

Am I doing something wrong? Is this a limitation of treesitter? How can I debug this?

**Reply by ahelwer** [reply]
I, too ran into this problem of requiring fine-grained control of lexical precedence while also breaking apart a token into fields. The easiest answer turned out to be using an external scanner, which gets highest lexical precedence by default. I used the external scanner to lookahead then emit invisible marker tokens like `_START_FOO_TOKEN` and `_END_FOO_TOKEN`, consumed in the grammar as follows:
```js
foo_token: $ => seq($._START_FOO_TOKEN, $.foo_token_inner_logic, $._END_FOO_TOKEN)
```
You can see my explorations here: https://github.com/tree-sitter/tree-sitter/discussions/1297

**Reply by ahlinc** [reply]
@BonaBeavis You don't doing anything wrong but the grammar, it seems, has enough of complexity and on the other hand EBNF notations don't precisely specify any lexical terminal conflicts, so in practice there is normal that concrete grammar implementation has many workarounds especially if EBNF notation was written for LL(k) grammar but implementation is LR grammar and vice versa.
I expected that your case in this discussion should be resolved by using the `word` token by seems it doesn't work and may be because actual case-insensitive `BASE` token is specified by a regex, I'm not sure does a keyword extraction feature works for regexes like it works for string literals.

> String tokens from the grammar seem always have precedence

Yes, this is a tree-sitter's feature, if you need to change such precedence you can use `token(prec(N, ...))` construction which specifies a lexer precedence while `prec(N, ...)` specifies a parser precedence for rules.

> How can I debug this?

Try to look into `tree-sitter parse -d` debug output, it may be unclear at the beginning and there no documentation how to interpret it, but actually it's very helpful and can explain what happens in around 80-90% of cases.

It's normal if you need to have nodes in terminals, @ahelwer suggested the right approach how to resolve conflicts on going this way.
_And feel free to ask questions by further concrete cases, we all here also learning something new with every new case._

**Reply by GordianDziwis** [reply]
> @BonaBeavis You don't doing anything wrong but the grammar, it seems, has enough of complexity and on the other hand EBNF notations don't precisely specify any lexical terminal conflicts, so in practice there is normal that concrete grammar implementation has many workarounds especially if EBNF notation was written for LL(k) grammar but implementation is LR grammar and vice versa.

Exactly this is the case: "The Turtle grammar is LL(1) and LALR(1) when the rules with uppercased names are used as terminals." 
This explains a lot, so implementing a grammar is in this case is always a bit of "art".

> Yes, this is a tree-sitter's feature, if you need to change such precedence you can use `token(prec(N, ...))` construction which specifies a lexer precedence while `prec(N, ...)` specifies a parser precedence for rules.

This and a bit of Trial and Error helped.

> It's normal if you need to have nodes in terminals, @ahelwer suggested the right approach how to resolve conflicts on going this way.
> _And feel free to ask questions by further concrete cases, we all here also learning something new with every new case._

Fortunately I got it to work without an external scanner. But I have the same Issues with the sparql grammar, which is a superset of turtle and  much mightier beast. I am happy to hear you are interested in concrete case, I will bring some forward.
